In [96]:
import pandas as pd
import numpy as np 

In [134]:
movie = pd.read_csv("tmdb_5000_movies.csv.zip")

In [135]:
credit = pd.read_csv("tmdb_5000_credits.csv.zip")

In [136]:
movie.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [137]:
credit.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

#fulljoin

In [138]:
movie = movie.merge(credit,on="title")

In [139]:
movie.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  ... runtime  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title vote_average vote_count movie_id  \
0  Enter the World of Pandora.  Avatar          7.2      11800    19995   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  

[1 rows x 23 columns]

In [140]:
movie.shape

(4809, 23)

#removing insufficient columns, the sufficient columns have should be contain tags to predict

In [141]:
movie["original_language"].value_counts()

original_language
en    4510
fr      70
es      32
zh      27
de      27
hi      19
ja      16
it      14
ko      12
cn      12
ru      11
pt       9
da       7
sv       5
nl       4
fa       4
th       3
he       3
ta       2
cs       2
ro       2
id       2
ar       2
vi       1
sl       1
ps       1
no       1
ky       1
hu       1
pl       1
af       1
nb       1
tr       1
is       1
xx       1
te       1
el       1
Name: count, dtype: int64

In [142]:
df = movie

In [143]:
columns_to_check = ['budget', 'genres', 'homepage', 'id', 'keywords', 'original_language', 'original_title', 'overview', 'popularity', 'production_companies', 'runtime', 'spoken_languages', 'status', 'tagline', 'title', 'vote_average', 'vote_count', 'movie_id', 'cast', 'crew']

# Part 1:Basic information for each column
print("=== PART 1 ===")
for column in columns_to_check[:5]: 
    print(f"--- {column.upper()} ---")
    print("Sample Data: ", df[column].head())  
    print("Unique Values: ", df[column].nunique())  
    print("\n")

# Part 2:Next 5 columns
print("=== PART 2 ===")
for column in columns_to_check[5:10]:
    print(f"--- {column.upper()} ---")
    print("Sample Data: ", df[column].head())  
    print("Unique Values: ", df[column].nunique())  # Unique values
    print("\n")

# Part 3:Next 5 columns
print("=== PART 3 ===")
for column in columns_to_check[10:15]:  
    print(f"--- {column.upper()} ---")
    print("Sample Data: ", df[column].head()) 
    print("Unique Values: ", df[column].nunique())  # Unique values
    print("\n")

# Part 4:Remaining columns
print("=== PART 4 ===")
for column in columns_to_check[15:]:  
    print(f"--- {column.upper()} ---")
    print("Sample Data: ", df[column].head())  
    print("Unique Values: ", df[column].nunique()) 
    print("\n")

=== PART 1 ===
--- BUDGET ---
Sample Data:  0    237000000
1    300000000
2    245000000
3    250000000
4    260000000
Name: budget, dtype: int64
Unique Values:  436


--- GENRES ---
Sample Data:  0    [{"id": 28, "name": "Action"}, {"id": 12, "nam...
1    [{"id": 12, "name": "Adventure"}, {"id": 14, "...
2    [{"id": 28, "name": "Action"}, {"id": 12, "nam...
3    [{"id": 28, "name": "Action"}, {"id": 80, "nam...
4    [{"id": 28, "name": "Action"}, {"id": 12, "nam...
Name: genres, dtype: object
Unique Values:  1175


--- HOMEPAGE ---
Sample Data:  0                     http://www.avatarmovie.com/
1    http://disney.go.com/disneypictures/pirates/
2     http://www.sonypictures.com/movies/spectre/
3              http://www.thedarkknightrises.com/
4            http://movies.disney.com/john-carter
Name: homepage, dtype: object
Unique Values:  1691


--- ID ---
Sample Data:  0     19995
1       285
2    206647
3     49026
4     49529
Name: id, dtype: int64
Unique Values:  4803


--- KEYWORDS

In [144]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

#Genres: Shows the type of movie, helping recommend similar ones.

#Movie ID: A unique number for each movie, necessary to track them.

#Keywords: Important words describing the movie, useful for better recommendations.

#Title: The movie's name, needed to identify and link the movies.

#Overview: A short description of the movie, helping to find similar ones.

#Cast: Actors in the movie, useful to recommend based on what users like.

#Crew: Directors and key staff, helps in finding similar movies.

In [145]:
# Keep only the selected columns
selected_columns = ['genres', 'movie_id', 'keywords', 'title', 'overview', 'cast', 'crew']

df= df[selected_columns]

In [146]:
df.head(5)

genres  movie_id  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...     19995   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...       285   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...    206647   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...     49026   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...     49529   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                      title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

#We will create three columns: movie id, title, and a new column called "tags." In the "tags" column, we will include all the information except movie id and title.

Take the overview and add genre values

For keywords, include only the key names, like "culture clash."

Add the top 3 actors from the cast.

Add the director's name from the crew.

Combine all of this into a paragraph for each movie.

In [147]:
df.isnull().sum()

genres      0
movie_id    0
keywords    0
title       0
overview    3
cast        0
crew        0
dtype: int64

In [148]:
df.dropna(inplace=True)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\1379821321.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.dropna(inplace=True)


In [149]:
df.isnull().sum()

genres      0
movie_id    0
keywords    0
title       0
overview    0
cast        0
crew        0
dtype: int64

In [150]:
df.duplicated().sum()

0

In [151]:
df["genres"][0]

'[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

#[{"id": 28, "name": "Action"}, {"id": 12, "name": "Adventure"}, {"id": 14, "name": "Fantasy"}, {"id": 878, "name": "Science Fiction"}]'

#frist we should have to convert upper list or dict into lower on list 

#["action","adventure","fantasy","Science Fiction"]

In [152]:
import ast
def convert(obj):
    l = []
    for i in ast.literal_eval(obj):
        l.append(i["name"])
    return l

In [153]:
df["genres"] = df["genres"].apply(convert)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\4187049850.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["genres"] = df["genres"].apply(convert)


In [154]:
df["genres"]

0       [Action, Adventure, Fantasy, Science Fiction]
1                        [Adventure, Fantasy, Action]
2                          [Action, Adventure, Crime]
3                    [Action, Crime, Drama, Thriller]
4                [Action, Adventure, Science Fiction]
                            ...                      
4804                        [Action, Crime, Thriller]
4805                                [Comedy, Romance]
4806               [Comedy, Drama, Romance, TV Movie]
4807                                               []
4808                                    [Documentary]
Name: genres, Length: 4806, dtype: object

In [155]:
df["genres"][0]

['Action', 'Adventure', 'Fantasy', 'Science Fiction']

In [156]:
df.head(1)

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   

                                            keywords   title  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...  Avatar   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

#this same function also on keywords column 

In [157]:
df["keywords"] = df["keywords"].apply(convert)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\2178206533.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["keywords"] = df["keywords"].apply(convert)


In [158]:
df["keywords"][0]

['culture clash',
 'future',
 'space war',
 'space colony',
 'society',
 'space travel',
 'futuristic',
 'romance',
 'space',
 'alien',
 'tribe',
 'alien planet',
 'cgi',
 'marine',
 'soldier',
 'battle',
 'love affair',
 'anti war',
 'power relations',
 'mind and soul',
 '3d']

In [159]:
df.sample(1)

genres  movie_id      keywords    title  \
4151  [Action, Adventure, Drama]     18808  [road movie]  Highway   

                                               overview  \
4151  Jack is caught with the wife of his employer, ...   

                                                   cast  \
4151  [{"cast_id": 3, "character": "Pilot Kelson", "...   

                                                   crew  
4151  [{"credit_id": "52fe479d9251416c7509fae9", "de...

In [160]:
df["cast"][0]

'[{"cast_id": 242, "character": "Jake Sully", "credit_id": "5602a8a7c3a3685532001c9a", "gender": 2, "id": 65731, "name": "Sam Worthington", "order": 0}, {"cast_id": 3, "character": "Neytiri", "credit_id": "52fe48009251416c750ac9cb", "gender": 1, "id": 8691, "name": "Zoe Saldana", "order": 1}, {"cast_id": 25, "character": "Dr. Grace Augustine", "credit_id": "52fe48009251416c750aca39", "gender": 1, "id": 10205, "name": "Sigourney Weaver", "order": 2}, {"cast_id": 4, "character": "Col. Quaritch", "credit_id": "52fe48009251416c750ac9cf", "gender": 2, "id": 32747, "name": "Stephen Lang", "order": 3}, {"cast_id": 5, "character": "Trudy Chacon", "credit_id": "52fe48009251416c750ac9d3", "gender": 1, "id": 17647, "name": "Michelle Rodriguez", "order": 4}, {"cast_id": 8, "character": "Selfridge", "credit_id": "52fe48009251416c750ac9e1", "gender": 2, "id": 1771, "name": "Giovanni Ribisi", "order": 5}, {"cast_id": 7, "character": "Norm Spellman", "credit_id": "52fe48009251416c750ac9dd", "gender": 

#want to extract frist three dict or there names for every movie

In [161]:
df['cast'] = df['cast'].apply(lambda x: [i['name'] for i in ast.literal_eval(x)[:3]])
print(df)

                                             genres  movie_id  \
0     [Action, Adventure, Fantasy, Science Fiction]     19995   
1                      [Adventure, Fantasy, Action]       285   
2                        [Action, Adventure, Crime]    206647   
3                  [Action, Crime, Drama, Thriller]     49026   
4              [Action, Adventure, Science Fiction]     49529   
...                                             ...       ...   
4804                      [Action, Crime, Thriller]      9367   
4805                              [Comedy, Romance]     72766   
4806             [Comedy, Drama, Romance, TV Movie]    231617   
4807                                             []    126186   
4808                                  [Documentary]     25975   

                                               keywords  \
0     [culture clash, future, space war, space colon...   
1     [ocean, drug abuse, exotic island, east india ...   
2     [spy, based on novel, secret agent, 

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\275010305.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['cast'] = df['cast'].apply(lambda x: [i['name'] for i in ast.literal_eval(x)[:3]])


In [162]:
df.head(1)

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [163]:
df["cast"]

0        [Sam Worthington, Zoe Saldana, Sigourney Weaver]
1           [Johnny Depp, Orlando Bloom, Keira Knightley]
2            [Daniel Craig, Christoph Waltz, Léa Seydoux]
3            [Christian Bale, Michael Caine, Gary Oldman]
4          [Taylor Kitsch, Lynn Collins, Samantha Morton]
                              ...                        
4804    [Carlos Gallardo, Jaime de Hoyos, Peter Marqua...
4805         [Edward Burns, Kerry Bishé, Marsha Dietlein]
4806           [Eric Mabius, Kristin Booth, Crystal Lowe]
4807            [Daniel Henney, Eliza Coupe, Bill Paxton]
4808    [Drew Barrymore, Brian Herzlinger, Corey Feldman]
Name: cast, Length: 4806, dtype: object

In [164]:
df["crew"][0]

'[{"credit_id": "52fe48009251416c750aca23", "department": "Editing", "gender": 0, "id": 1721, "job": "Editor", "name": "Stephen E. Rivkin"}, {"credit_id": "539c47ecc3a36810e3001f87", "department": "Art", "gender": 2, "id": 496, "job": "Production Design", "name": "Rick Carter"}, {"credit_id": "54491c89c3a3680fb4001cf7", "department": "Sound", "gender": 0, "id": 900, "job": "Sound Designer", "name": "Christopher Boyes"}, {"credit_id": "54491cb70e0a267480001bd0", "department": "Sound", "gender": 0, "id": 900, "job": "Supervising Sound Editor", "name": "Christopher Boyes"}, {"credit_id": "539c4a4cc3a36810c9002101", "department": "Production", "gender": 1, "id": 1262, "job": "Casting", "name": "Mali Finn"}, {"credit_id": "5544ee3b925141499f0008fc", "department": "Sound", "gender": 2, "id": 1729, "job": "Original Music Composer", "name": "James Horner"}, {"credit_id": "52fe48009251416c750ac9c3", "department": "Directing", "gender": 2, "id": 2710, "job": "Director", "name": "James Cameron"},

#From each dictionary,need the dictionary where the job value is "Director," and then extract the name from that dictionary

In [165]:
df['crew'] = df['crew'].apply(lambda crew: [person['name'] for person in eval(crew) if isinstance(person, dict) and person.get('job') == 'Director'] if isinstance(crew, str) else [])

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3233946351.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['crew'] = df['crew'].apply(lambda crew: [person['name'] for person in eval(crew) if isinstance(person, dict) and person.get('job') == 'Director'] if isinstance(crew, str) else [])


In [166]:
df.head(1)

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

#we convert the overview "str" into list for merge columns expected movie_id and title

In [167]:
df["overview"][0]

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

In [168]:
df["overview"] = df["overview"].apply(lambda x:x.split())

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\2705624388.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["overview"] = df["overview"].apply(lambda x:x.split())


In [169]:
df.head(1)

genres  movie_id  \
0  [Action, Adventure, Fantasy, Science Fiction]     19995   

                                            keywords   title  \
0  [culture clash, future, space war, space colon...  Avatar   

                                            overview  \
0  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

#If there are spaces, when matching, movies with similar names or words might be wrongly recommended.

In [170]:
columns_to_clean = ["genres", "keywords", "cast", "crew"]
df[columns_to_clean] = df[columns_to_clean].applymap(lambda x: [i.replace(" ", "") for i in x] if isinstance(x, list) else x)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3885569577.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df[columns_to_clean] = df[columns_to_clean].applymap(lambda x: [i.replace(" ", "") for i in x] if isinstance(x, list) else x)
C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3885569577.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[columns_to_clean] = df[columns_to_clean].applymap(lambda x: [i.replace(" ", "") for i in x] if isinstance(x, list) else x)


In [171]:
df.sample(1)

genres  movie_id  \
2511  [Comedy, Horror, ScienceFiction]      9035   

                                               keywords    title  \
2511  [smalltown, mutant, meteor, meat, alien, viole...  Slither   

                                               overview  \
2511  [A, small, town, is, taken, over, by, an, alie...   

                                              cast         crew  
2511  [NathanFillion, DonThompson, ElizabethBanks]  [JamesGunn]

#create a tags name column of combination or ["genres", "keywords", "cast", "crew"]

In [172]:
df['tags'] = df['genres'] + df['keywords'] + df['overview'] + df['cast'] + df['crew']

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3562865596.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tags'] = df['genres'] + df['keywords'] + df['overview'] + df['cast'] + df['crew']


In [173]:
df.sample(1)

genres  movie_id  \
618  [Adventure, Fantasy, Action, Comedy, ScienceFi...      9824   

                                              keywords        title  \
618  [bowling, hostage, sphinx, training, insaneasy...  Mystery Men   

                                              overview  \
618  [When, Captain, Amazing, (Kinnear), is, kidnap...   

                                          cast          crew  \
618  [BenStiller, GeoffreyRush, ClaireForlani]  [KinkaUsher]   

                                                  tags  
618  [Adventure, Fantasy, Action, Comedy, ScienceFi...

In [174]:
df.drop(['genres', 'keywords', 'overview', 'cast', 'crew'], axis=1, inplace=True)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3778837553.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(['genres', 'keywords', 'overview', 'cast', 'crew'], axis=1, inplace=True)


In [176]:
df.head(5)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [Action, Adventure, Fantasy, ScienceFiction, c...  
1  [Adventure, Fantasy, Action, ocean, drugabuse,...  
2  [Action, Adventure, Crime, spy, basedonnovel, ...  
3  [Action, Crime, Drama, Thriller, dccomics, cri...  
4  [Action, Adventure, ScienceFiction, basedonnov...

#now convert tags list format into string because of matching to recommend

In [177]:
df['tags'] = df['tags'].apply(lambda x: ' '.join(x))

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\1753472850.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['tags'] = df['tags'].apply(lambda x: ' '.join(x))


In [178]:
df.sample(5)

movie_id                         title  \
1446     18320            The Young Victoria   
4221      3083  Mr. Smith Goes to Washington   
823       8963                  Crimson Tide   
1720     10439                   Hocus Pocus   
4065    297621      There's Always Woodstock   

                                                   tags  
1446  Drama History Romance royalfamily biography ro...  
4221  Comedy Drama corruption washingtond.c. senate ...  
823   Action Thriller Drama submarine mutiny russia ...  
1720  Comedy Family Fantasy witch halloween salem tr...  
4065  Romance Comedy Music womandirector When Neurot...

In [179]:
df["tags"][0]

'Action Adventure Fantasy ScienceFiction cultureclash future spacewar spacecolony society spacetravel futuristic romance space alien tribe alienplanet cgi marine soldier battle loveaffair antiwar powerrelations mindandsoul 3d In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization. SamWorthington ZoeSaldana SigourneyWeaver JamesCameron'

In [180]:
df["tags"] = df["tags"].apply(lambda x:x.lower())

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\3213508868.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["tags"] = df["tags"].apply(lambda x:x.lower())


#form the basies of tags column how can we recommend the similar moive 
#And how we can create similarity between 2 or more moive tag value because they are text not numberic

#text Vectorization is the solution 

#We will perform vectorization on the tags column. For this, we first combine all the tag values into one large string. This method is called Bag of Words, where we merge all the text from the tags column into a single string. 
#Then, we identify the most common words from this combined text.

#Next, we decide how many most common words to consider. For now, we are taking 5000 frequent words. Taking more than this might slow the system because it increases dimensionality and makes the process less efficient.

#After finding the most common words, we remove stop words like is, am, are, and, the, of, from, etc., as they are not useful for recommendations.

#Once we have the most common words, we check how many times each of these words appear in the text for each movie’s tags. This process is repeated for all movies.

#We then create a matrix of counts, where each row represents a movie and each column represents the count of a specific common word in that movie's tags. 
#This gives us a vector for each movie based on the count of the most common words.

#Finally, we calculate the similarity between these vectors. 
#If a movie (say X) has a vector close to other movies, we recommend its 5 closest movies. These similar vectors indicate that the movies have related content, making them suitable recommendations.

In [181]:
df.head(1)

movie_id   title                                               tags
0     19995  Avatar  action adventure fantasy sciencefiction cultur...

#where is any words be like [loving , loved , love] so these are the words mean same for this apply steming to overcome 

In [182]:
import nltk 
from nltk.stem.porter import PorterStemmer
ps = PorterStemmer()

In [183]:
def stem(text):
    y = []
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [184]:
df["tags"] = df["tags"].apply(stem)

C:\Users\HP\AppData\Local\Temp\ipykernel_6552\552104661.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["tags"] = df["tags"].apply(stem)


In [185]:
df.sample(5)

movie_id                 title  \
1887    152737  August: Osage County   
4402     91122            16 to Life   
3723    281730                 Fugly   
2375     10549                Hamlet   
4544     13983    Finishing The Game   

                                                   tags  
1887  comedi drama suicid drugaddict funer dysfuncti...  
4402  comedi romanc affect womandirector romant come...  
3723  comedi drama thriller it is a stori of 4 frien...  
2375  drama mother shakespear denmark lossoffath pri...  
4544  comedi satir independentfilm spoof in 1973, ma...

#process of making vactor

In [186]:
from sklearn.feature_extraction.text import CountVectorizer

In [187]:
cv = CountVectorizer(max_features=5000 , stop_words="english")
vectors = cv.fit_transform(df["tags"]).toarray()

In [188]:
vectors

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

In [197]:
vectors.shape

(4806, 5000)

In [189]:
vectors[0]

array([0, 0, 0, ..., 0, 0, 0], dtype=int64)

In [190]:
cv.get_feature_names_out()

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

#We calculate the distance between each vector using cosine similarity, not Euclidean distance. This is because Euclidean distance often fails in high-dimensional spaces. Cosine similarity works better as it measures the angle between vectors. A smaller angle means higher similarity, making it more effective for our use case.

In [191]:
from sklearn.metrics.pairwise import cosine_similarity

In [198]:
similarity = cosine_similarity(vectors)
similarity

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

In [200]:
similarity.shape

(4806, 4806)

In [199]:
similarity[0] #this similarity shows the distance between all vectors

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

In [194]:
similarity[1]

array([0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
       0.02615329])

#now the code which recommend 5 similar moives 

In [195]:
def recommend(movie):
    movie_index = df[df["title"] == movie].index[0]
    distances = similarity[movie_index]
    movies_list = sorted(list(enumerate(distances)),reverse=True,key=lambda x:x[1])[1:6]
    
    for i in movies_list:
        print(df.iloc[i[0]].title)

In [196]:
recommend("Avatar")

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.
